### PR TITLE
feat: client validation

### PR DESCRIPTION
**Pull request by QuentinBrosse**
The goal here is to have a basic but robust validation of the data given for the creation of a client.

We expose a `ClientValidationError` error that can be cached by users.

A more precise validation can be done by the user before the `NewClient` call ([example](https://github.com/scaleway/scaleway-sdk-go/pull/234#issue-340526694)).

### CODE REVIEW

File: scw/client_option.go [hunk 1]
```diff
@@ -170,40 +171,70 @@ func (s *settings) apply(opts []ClientOption) {
...
 		panic(errors.New("no credential option provided"))
 	}
-
 	if token, isToken := s.token.(*auth.Token); isToken {
 		if token.AccessKey == "" {
-			return &ClientCredentialError{errorType: clientCredentialError_EmptyAccessKey}
+			return NewClientValidationError("access key cannot be empty")
+		}
+		if !validation.IsAccessKey(token.AccessKey) {
+			return NewClientValidationError("bad access key format")
```
Comment (kindermoumoute) [inline scw/client_option.go]: Give a hint: `access key format must be: SCWXXXXXX`

File: scw/client_option.go [hunk 2]
```diff
@@ -170,40 +171,70 @@ func (s *settings) apply(opts []ClientOption) {
...
 		}
 		if token.SecretKey == "" {
-			return &ClientCredentialError{errorType: clientCredentialError_EmptySecreyKey}
+			return NewClientValidationError("secret key cannot be empty")
+		}
+		if !validation.IsSecretKey(token.SecretKey) {
+			return NewClientValidationError("bad secret key format")
```
Comment (kindermoumoute) [inline scw/client_option.go]: Give a hint: `secret key format must be a valid UUID: xxx-xx-x`

File: scw/client_option.go [hunk 3]
```diff
@@ -170,40 +171,70 @@ func (s *settings) apply(opts []ClientOption) {
...
+	if s.defaultRegion != nil {
+		if *s.defaultRegion == "" {
+			return NewClientValidationError("default region cannot be empty")
+		}
+		if !validation.IsRegion(string(*s.defaultRegion)) {
+			regions := []string(nil)
+			for _, r := range AllRegions {
+				regions = append(regions, string(r))
+			}
+			return NewClientValidationError("bad default region format, available regions are: %s", strings.Join(regions, ", "))
```
Comment (kindermoumoute) [inline scw/client_option.go]: precise the value checked `*s.defaultRegion is not a valid region`

File: scw/client_option.go [hunk 4]
```diff
@@ -170,40 +171,70 @@ func (s *settings) apply(opts []ClientOption) {
...
-	_, err = url.Parse(s.apiURL)
-	if err != nil {
-		return errors.Wrap(err, "invalid url %s", s.apiURL)
+	// Default Organization ID.
+	if s.defaultOrganizationID != nil {
+		if *s.defaultOrganizationID == "" {
+			return NewClientValidationError("default organization ID cannot be empty")
+		}
+		if !validation.IsOrganizationID(*s.defaultOrganizationID) {
+			return NewClientValidationError("default organization ID must be a valid UUID")
```
Comment (kindermoumoute) [inline scw/client_option.go]: precise the value checked `*s. defaultOrganizationID`

File: scw/client_option.go (hunk 1 above)
Comment (kindermoumoute) [reply scw/client_option.go]: + precise the value checked `token.AccessKey`

File: scw/client_option.go (hunk 2 above)
Comment (kindermoumoute) [reply scw/client_option.go]: precise the anonymised value checked `1f3e9xxxxx-xxx-xxx-x`

File: scw/locality_test.go [hunk 5]
```diff
@@ -34,17 +34,17 @@ func TestParseZone(t *testing.T) {
 		{
 			input:    "fr-par",
 			expected: "",
-			err:      errors.New("wrong zone format, available zones are: fr-par-1, fr-par-2, nl-ams-1"),
+			err:      errors.New("bad zone format, available zones are: fr-par-1, fr-par-2, nl-ams-1"),
```
Comment (kindermoumoute) [inline scw/locality_test.go]: invalid? 🤔

File: scw/errors.go [hunk 6]
```diff
@@ -256,20 +256,17 @@ func (e *OutOfStockError) GetRawBody() json.RawMessage {
...
 }
 
-type clientCredentialErrorType string
-
-const (
-	clientCredentialError_EmptyAccessKey = clientCredentialErrorType("access key cannot be empty")
-	clientCredentialError_EmptySecreyKey = clientCredentialErrorType("secret key cannot be empty")
-)
+// ClientValidationError indicates that at least one of client data have been badly provided for the client creation.
+type ClientValidationError struct {
```
Comment (jerome-quere) [inline scw/errors.go]: what about `InvalidClientOptionError` ?

File: internal/validation/is.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"net/url"
+	"regexp"
+)
+
+var (
+	isUUIDRegexp  = regexp.MustCompile(`[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}`)
+	isRegionRegex = regexp.MustCompile("^[a-z]{2}-[a-z]{3}$")
+	isZoneRegex   = regexp.MustCompile("^[a-z]{2}-[a-z]{3}-[1-9]$")
```
Comment (remyleone) [inline internal/validation/is.go]: Are uppercase regions/zones valid for API calls?

File: scw/client_option.go (hunk 3 above)
Comment (QuentinBrosse) [reply scw/client_option.go]: `*s. defaultOrganizationID` is an internal structure, it brings no value for the final user.

File: scw/errors.go (hunk 6 above)
Comment (QuentinBrosse) [reply scw/errors.go]: "Invalid" and "Error" in the same name? 🤔

File: internal/validation/is.go (hunk 7 above)
Comment (QuentinBrosse) [reply internal/validation/is.go]: nop

File: scw/client_option.go [hunk 8]
```diff
@@ -181,13 +181,13 @@ func (s *settings) validate() error {
...
 		if !validation.IsAccessKey(token.AccessKey) {
-			return NewClientValidationError("bad access key format")
+			return NewClientValidationError("invalid access key format '%s', must be: SCWXXXXXXXXXXXXXXXXX", token.AccessKey)
 		}
 		if token.SecretKey == "" {
 			return NewClientValidationError("secret key cannot be empty")
 		}
 		if !validation.IsSecretKey(token.SecretKey) {
-			return NewClientValidationError("bad secret key format")
+			return NewClientValidationError("invalid access key format '%s', must be: SCWXXXXXXXXXXXXXXXXX", token.AccessKey)
```
Comment (kindermoumoute) [inline scw/client_option.go]: nit: `"invalid access key format '%s': expected SCWXXXXXXXXXXXXXXXXX format"`

File: scw/client_option.go [hunk 9]
```diff
@@ -170,40 +171,70 @@ func (s *settings) apply(opts []ClientOption) {
...
-	_, err = url.Parse(s.apiURL)
-	if err != nil {
-		return errors.Wrap(err, "invalid url %s", s.apiURL)
+	// Default Organization ID.
+	if s.defaultOrganizationID != nil {
+		if *s.defaultOrganizationID == "" {
+			return NewClientValidationError("default organization ID cannot be empty")
+		}
+		if !validation.IsOrganizationID(*s.defaultOrganizationID) {
+			return NewClientValidationError("invalid organization ID format '%s', must be an UUID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", *s.defaultOrganizationID)
```
Comment (kindermoumoute) [inline scw/client_option.go]: `a UUID`*

File: scw/errors.go (hunk 6 above)
Comment (jerome-quere) [reply scw/errors.go]: Yes : Client option is Invalid  and this is an error

In Java or PHP class name are `InvalidParameterException` for example

**Pull request merged**